ReadersAndroidJetpack/app/src/main/java/com/kontinua/readersandroidjetpack/viewmodels/BookmarkViewModel.kt

```diff
@@ -0,0 +1,55 @@
+package com.kontinua.readersandroidjetpack.viewmodels
+
+import android.app.Application
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.viewModelScope
+import com.kontinua.readersandroidjetpack.data.BookmarkRepository
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+class BookmarkViewModel(application: Application) : AndroidViewModel(application) {
+    private val repository = BookmarkRepository(application.applicationContext)
+
+    private val _bookmarkLookup = MutableStateFlow<Map<Int, Set<Int>>>(emptyMap())
+    val bookmarkLookup: StateFlow<Map<Int, Set<Int>>> = _bookmarkLookup.asStateFlow()
+
+    init {
+        viewModelScope.launch {
+            // Load initial bookmarks from DataStore and keep observing
+            repository.bookmarkLookupFlow.collect { loadedBookmarks ->
+                _bookmarkLookup.value = loadedBookmarks
+            }
+        }
+    }
+
+//    fun isBookmarked(workbookId: Int, currentPage: Int): Boolean {
+//        return _bookmarkLookup.value[workbookId]?.contains(currentPage) ?: false
+//    }
+
+    fun toggleBookmark(workbookId: Int, currentPage: Int) {
+        _bookmarkLookup.update { currentLookup ->
+            val mutableLookup = currentLookup.toMutableMap()
+            val pagesForWorkbook = mutableLookup[workbookId]?.toMutableSet() ?: mutableSetOf()
+
+            if (pagesForWorkbook.contains(currentPage)) {
```

Comment:
The verbosity of this common operation bugged me. You can actually boil it down to a oneliner thanks to add and remove returning true if they changed the set and false if they wound up being a no-op, by trying first one then the other:

```kotlin
fun <T> MutableSet<T>.toggle(v: T) = this.add(v) || this.remove(v)
```